universe/base.go

```diff
@@ -223,34 +246,64 @@ func (a *MintingArchive) RegisterIssuance(ctx context.Context, id Identifier,
 
 	// The script key should also match exactly.
 	case !newAsset.ScriptKey.PubKey.IsEqual(key.ScriptKey.PubKey):
-		return nil, fmt.Errorf("script key mismatch: expected %v, "+
-			"got %v", key.ScriptKey.PubKey.SerializeCompressed(),
+		return nil, fmt.Errorf("script key mismatch: expected %v, got "+
+			"%v", key.ScriptKey.PubKey.SerializeCompressed(),
 			newAsset.ScriptKey.PubKey.SerializeCompressed())
 	}
 
-	// Now that we know the proof is valid, we'll insert it into the base
-	// multiverse backend, and return the new issuance proof.
-	issuanceProof, err := a.cfg.Multiverse.RegisterIssuance(
-		ctx, id, key, leaf, assetSnapshot.MetaReveal,
+	return assetSnapshot, nil
+}
+
+// RegisterNewIssuanceBatch inserts a batch of new minting leaves within the
+// target universe tree (based on the ID), stored at the base key(s). We assume
+// the proofs within the batch have already been checked that they don't yet
+// exist in the local database.
+func (a *MintingArchive) RegisterNewIssuanceBatch(ctx context.Context,
```

Comment:
In `RegisterIssuance` we attempted to fetch each proof before insertion, and that's dropped in this batched version. IIUC the actual registration is idempotent, so the main consequence of not having that fetching here is re-verifying those proofs that may already be inserted?

Reply:
I would say the main difference is that `RegisterIssuance` is also the method that's being called by the RPC when a proof is sent to an universe. And we probably do want to have the fetch first to make sure we don't start an "expensive" write operation just to end up in an upsert that doesn't really change any data. That would be a DoS vector.
But since we _know_ all the proofs in a batch of leaves we are syncing from another universe, we kind of already know those should all be new. And `RegisterNewIssuanceBatch` isn't currently hooked up to an RPC so it's not at risk of being a DoS vector.